core/src/main/java/org/apache/iceberg/SnapshotManager.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.base.Preconditions;
+import java.util.HashSet;
+import java.util.Set;
+import org.apache.iceberg.exceptions.DuplicateWAPCommitException;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.util.SnapshotUtil;
+
+public class SnapshotManager extends MergingSnapshotProducer<ManageSnapshots> implements ManageSnapshots {
+
+  private SnapshotManagerOperation operation;
+  private TableMetadata base;
+  private Long targetSnapshotId = null;
+  private Set<Long> snapshotsAlreadyCherrypicked;
+
+  enum SnapshotManagerOperation {
+    CHERRYPICK,
+    ROLLBACK
+  }
+
+  SnapshotManager(TableOperations ops) {
+    super(ops);
+    this.base = ops.current();
+    this.snapshotsAlreadyCherrypicked = new HashSet<>();
+  }
+
+  @Override
+  protected ManageSnapshots self() {
+    return this;
+  }
+
+  @Override
+  protected String operation() {
+    Snapshot manageSnapshot = base.snapshot(targetSnapshotId);
+    return manageSnapshot.operation();
+  }
+
+  @Override
+  public ManageSnapshots cherrypick(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot cherry pick unknown snapshot id: %s", snapshotId);
+
+    operation = SnapshotManagerOperation.CHERRYPICK;
+    this.targetSnapshotId = snapshotId;
+
+    // Pick modifications from the snapshot
+    Snapshot cherryPickSnapshot = base.snapshot(this.targetSnapshotId);
+    String wapId = stagedWapId(cherryPickSnapshot);
+    boolean isWapWorkflow =  wapId != null && !"".equals(wapId);
+    if (isWapWorkflow) {
+      if (base.isWapIdPublished(wapId)) {
+        throw new DuplicateWAPCommitException(wapId);
+      }
+    }
+    // only append operations are currently supported
+    if (!cherryPickSnapshot.operation().equals(DataOperations.APPEND)) {
+      throw new UnsupportedOperationException("Can cherry pick only append operations");
+    }
+    // this is to ensure we add files only once for each targetSnapshotId, to protect from
+    // duplicate additions if commit retries the same cherrypick operation on failure.
```

Comment:
Is this still needed if calls to `add` are in the configuration method rather than in `apply`? I don't think so. The configuration method will be called just once and will queue up just one set of changes.

Reply:
+1